Add the spec file to generate RPM for Fedora and RHEL

spetrosi opened the pull request:
https://bugzilla.redhat.com/show_bug.cgi?id=1967954

spetrosi committed:
Hack the spec file to work with the mssql role

Define target_rolename and collection_rolename if not defined

Do not copy .ansible-lint because the server role passes all checks

Use galaxy.yml and README.md for the collection from mssql repo

Remove the collection readme script

nhosoi commented on ansible-collection-microsoft-sql/ansible-collection-microsoft-sql.spec:
```diff
@@ -227,27 +105,14 @@ BuildRequires: python3dist(ruamel.yaml)
 Requires: python3-jmespath
 Requires: python3-netaddr
```
I wonder these Requires are needed?
`jmespath` is required if the role uses `json_query`.
`netaddr` was introduced by `vpn`.

spetrosi replied:
Hi Noriko, thank you for the review.

I didn't know that, I will remove these requirements now. The mssql role does not have any requirements.

nhosoi commented on ansible-collection-microsoft-sql/ansible-collection-microsoft-sql.spec:
```diff
@@ -308,71 +140,54 @@ for rolename in %{rolenames}; do
+# Declare the array containing collection rolenames to convert roles to
+declare -A COLLECTION_ROLENAMES=(%{collection_rolenames})
 
-rm -f .collections/ansible_collections/%{collection_namespace}/%{collection_name}/tests/sanity/ignore-2.9.txt
```
I'm curious about the future of mssql family which is going to use this spec file. Since you keep the array of rolenames, I assume the family member size would increase. If the assumption is valid, it might be a good idea to keep the framework for `requirements.txt`, `bindep.txt`, and `ignore-#.#.txt` although the current mssql needs none of them.

spetrosi replied:
I want to keep the spec file as minimalistic as possible for now, now this framework is not needed so I think there is no need to keep it, it would only cause confusion. I keep in mind that this exists in the linux-system-roles spec file and will copy it to here if needed in future.

nhosoi commented on ansible-collection-microsoft-sql/ansible-collection-microsoft-sql.spec:
```diff
@@ -308,71 +140,54 @@ for rolename in %{rolenames}; do
+# Copy README.md for the collection build
+cp mssql/.collection/README.md lsr_role2collection/collection_readme.md
+# Copy galaxy.yml for the collection build
+cp mssql/.collection/galaxy.yml ./
```
Sorry, this comment is about outside of this pr. The galaxy.yml file is for the entire sql collection not just for the mssql | sql-server role. Is it reasonable to have this Description "[The Ansible role that manages Microsoft SQL Server](https://github.com/linux-system-roles/mssql/blob/master/.collection/galaxy.yml#L4)"? I.e., if other roles are added to `sql`, the `README.md` and `galaxy.yml` in `mssql/.collection` may need to be updated for the new roles.

spetrosi replied:
Here I think it is fine to be fair with users and say straight out what is stored inside the collection. If the collection's readme says "This collection stores content to manage the SQL server" to me it builds the expectation that there is more than one role. I want to update it in future if more content is added.

nhosoi replied:
![ah](https://user-images.githubusercontent.com/938485/121390792-9319ef00-c902-11eb-9f9d-61d768bab99a.png)
The description `Red Hat Enterprise Linux System Roles Ansible Collection` is displayed like this on the Automation Hub page. This wording was discussed including Terry, Brian and Peter. I suggest to have their inputs, as well. IMHO, the `title` is supposed to be generic and stays the same with the contents grow.

spetrosi replied:
Thank's for the details, @nhosoi. I've updated the description to a generic `Ansible collection for Microsoft SQL Server management`. By updated I mean that I changed it within the galaxy_transform.py script.

nhosoi replied:
Thank you, @spetrosi. Your new description looks good to me.

spetrosi committed:
Copy Noriko's changes from linux-system-roles to fix README.html loc

Copy changes from
https://src.fedoraproject.org/fork/nhosoi/rpms/linux-system-roles/c/72032a4fd8baf354ef05bee4f6a7586520d0b8d4

spetrosi committed:
Remove requirement specific to linux-system-roles

richm commented on ansible-collection-microsoft-sql/ansible-collection-microsoft-sql.spec:
```diff
@@ -308,162 +131,109 @@ for rolename in %{rolenames}; do
 done
-%endif
 
+# Copy README, COPYING, and LICENCE files to the corresponding directories
```
LICENSE

spetrosi replied:
Done.

richm commented on ansible-collection-microsoft-sql/ansible-collection-microsoft-sql.spec:
```diff
@@ -575,220 +343,5 @@ fi
-- network r. 03ff040df78a14409a0d89eba1235b8f3e50a750
-
+* Thu Jun 3 2021 Sergei Petrosian <spetrosi@redhat.com> - 0.0.1-1
+- Initiale release
```
Initial

spetrosi replied:
Thank you, done.

spetrosi committed:
Fix typos

spetrosi committed:
Use macro instead of the plain mssql for clarity

pcahyna commented on ansible-collection-microsoft-sql/ansible-collection-microsoft-sql.spec:
```diff
@@ -54,159 +35,50 @@ BuildRequires: ansible >= 2.9.10
+# We don't have ansible-galaxy.
 # Simply copy everything instead of galaxy-installing the built artifact.
-%define ansible_collection_install() mkdir -p %{buildroot}%{ansible_collection_files}/%{collection_name}; (cd %{buildroot}%{ansible_collection_files}/%{collection_name}; tar -xf %{_tmppath}/%{collection_namespace}-%{collection_name}-%{version}.tar.gz)
+%define ansible_collection_build_install() tar -cf %{_tmppath}/%{collection_namespace}-%{collection_name}-%{version}.tar.gz .; mkdir -p %{buildroot}%{ansible_collection_files}/%{collection_name}; (cd %{buildroot}%{ansible_collection_files}/%{collection_name}; tar -xf %{_tmppath}/%{collection_namespace}-%{collection_name}-%{version}.tar.gz)
```
I think you should follow the macros (`ansible_collection_install()`, `ansible_collection_build()`) from the Ansible package ( https://src.fedoraproject.org/rpms/ansible/blob/rawhide/f/macros.ansible ) instead of defining your own `ansible_collection_build_install()` which does not have, AFAIK, any prior art.

richm replied:
> I think you should follow the macros (`ansible_collection_install()`, `ansible_collection_build()`) from the Ansible package ( https://src.fedoraproject.org/rpms/ansible/blob/rawhide/f/macros.ansible ) instead of defining your own `ansible_collection_build_install()` which does not have, AFAIK, any prior art.

it does have prior art - https://src.fedoraproject.org/rpms/linux-system-roles/blob/rawhide/f/linux-system-roles.spec#_59

pcahyna replied:
Ah. I missed this change, I would not expect it in a PR called "Fix README rendering problems".

pcahyna commented on ansible-collection-microsoft-sql/ansible-collection-microsoft-sql.spec:
```diff
@@ -54,159 +35,50 @@ BuildRequires: ansible >= 2.9.10
 # list of assignments that can be used to populate a bash associative array variable
 %global rolestodir %nil
+# list of target rolenames to copy the roles to
+%global target_rolenames %nil
```
can't you use `%rolenameN` for the target name and `%repoN` for the original name? Not sure whether this will work, but some kind of renaming is already done for the sshd role this way (the repo is called `ansible-sshd` and we call the role `linux-system-roles.sshd`).

spetrosi replied:
Hi Pavel, thank you for the review. Indeed I now see that the original linux-system-roles spec file had a solution to this and I overviewed it.
I have tried to implement this and faced an issue. To rename the role during the conversion to the collection, the original role must have a corresponding name. After I used the solution you proposed, rpmbuild extracts the `mssql` role and copies it to the `sql-server` directory. Then, it runs the role2collection script like that (pasting the parsed command from the buildlog for readability):
```
python3 lsr_role2collection.py --role sql-server \
--src-path sql-server \
--dest-path .collections \
--readme lsr_role2collection/collection_readme.md \
--namespace microsoft \
--collection sql \
--new-role server
```
Because the `--new-role` is set to `server`, the script tries to rename the role from `sql-server` to `server`, however, the role in fact uses the `mssql` wording everywhere. So the issue is that the scrip tries to rename e.g. `linux-system-roles.sql-server` to `microsoft.sql.server`, which is incorrect because in fact the role is called as `linux-system-roles.mssql`.

I tried to use the `--extra-mapping` option that @nhosoi added to lsr_role2collection.py instead of `--new-role` like `--extra-mapping mssql:server` but it didn't work for me. I suppose this option does something else. Noriko, can you justify?

spetrosi replied:
Apparently, to run this lsr_role2collection.py script the spec file will need to declare an array containing the source name and the target name, exactly what `TARGET_NAMES` does. So after all I think there is no need to re-use the `repo` macro, rather get rid of it as I have `target_namesN` for that. And after all, mssql is not only the repo name but the actual name of the source role.

spetrosi replied:
@pcahyna so using the existing `repo` macro would require the same functionality as the `target_rolenames` macro does, and I find the `target_rolename` macro names more descriptive than `repo`, so I kept `target_rolename` and got rid of unused `repo` and `forge` macros. Does this sound fine to you?

pcahyna replied:
well `repo` is not unused, it is used to construct the complete URL (it is the last part of the complete URL of the repo). Sure, you can remove it and construct the URL in a different way, but then your spec file will become a bit different from what is used in linux-system-roles. Since linux-system-roles could also use this kind of renaming mechanics, I think it would be better to share it, but at the end it is up to you.

spetrosi replied:
I think I am not going to keep the two spec files as identical as possible. They are too different. The reasoning behind keeping them identical and e.g. keeping unused macros would be to enable easy copying of changes from linux-system-roles.spec to ansible-collection-microsoft-sql.spec. I do not think that this will be the case because the files are already much different and are used differently. So picking changes from linux-system-roles anyway will be a manual process of reviewing each change and decing if it is needed or not.
So I think here I am trying to create a new spec file basing on the linux-system-roles.spec. This spec file will always be smaller and simpler.

richm replied:
ok - but if you figure out a better way to do something in linux-system-roles.spec, please submit a PR

nhosoi replied:
@spetrosi, sorry for my slow response. I think you already reached the conclusion, but just in case...

> I tried to use the --extra-mapping option that @nhosoi added to lsr_role2collection.py instead of --new-role like --extra-mapping mssql:server but it didn't work for me. I suppose this option does something else. Noriko, can you justify?

That's correct. --extra-mapping does not step on toes of --new-role to avoid the ambiguity.

spetrosi committed:
Point to the src-owner to perform correct renaming when converting

spetrosi committed:
Remove unused repo and forgeorg macros

richm commented on ansible-collection-microsoft-sql/ansible-collection-microsoft-sql.spec:
```diff
@@ -54,159 +35,49 @@ BuildRequires: ansible >= 2.9.10
+
 %define getarchivedir() %(p=%{basename:%{S:%{1}}}; echo ${p%%.%{archiveext}})
 
+%global parenturl https://github.com/linux-system-roles/
```
I think it is more natural to have no slash `/` at the end of the url string here, and instead where you use it, you use `/`

richm replied:
and yes, I realize that linux-system-roles has the `/` at the end of the url

spetrosi replied:
I've applied this improvement here and opened a PR for linux-system-roles too, please take a look if interested:
https://src.fedoraproject.org/fork/spetros/rpms/linux-system-roles/diff/rawhide..naturalize-url

spetrosi replied:
Correct URL to the PR: https://src.fedoraproject.org/rpms/linux-system-roles/pull-request/14

richm commented on ansible-collection-microsoft-sql/ansible-collection-microsoft-sql.spec:
```diff
@@ -54,159 +35,49 @@ BuildRequires: ansible >= 2.9.10
 
 %global mainid 17baba973c8c77b32ae65838a98dc4fbaf7b7b3e
-Source: %{url}auto-maintenance/archive/%{mainid}/auto-maintenance-%{mainid}.tar.gz
+Source: %{parenturl}auto-maintenance/archive/%{mainid}/auto-maintenance-%{mainid}.tar.gz
```
e.g.
```
Source: %{parenturl}/auto-maintenance/archive/%{mainid}/auto-maintenance-%{mainid}.tar.gz
```

spetrosi committed:
Naturalize urls

It is more natural to have no slash / at the end of the url definition,
and instead use / where the url is used

spetrosi committed:
Fix the collection description to be more generic

richm reviewed (approved):
lgtm

spetrosi commented:
I need to run `fedpkg request-repo ansible-collection-microsoft-sql 1967954` pointing to the Bugzilla to request a repo for this RPM In Fedora. I do not know exactly what are the requirements for it, but since the `fedpkg request-repo` requires only a name and a BZ number, a BZ approved by packagers might be enough. Please approve here:

https://bugzilla.redhat.com/show_bug.cgi?id=1967954

spetrosi committed:
Update summary and description fields to be more informative

spetrosi commented:
The spec file has been reviewed and is now released at https://src.fedoraproject.org/rpms/ansible-collection-microsoft-sql/blob/rawhide/f/ansible-collection-microsoft-sql.spec
